### PR TITLE
Move `calc_envelope` to LaserPulse and add tests

### PR DESCRIPTION
**Pull request by danielhollas**
Moved `calc_envelope` method to `LaserPulse` class. Also moved the `tmin` and `tmin` attributes and their computation into `LaserPulse`.

### CODE REVIEW

File: src/promdens/promdens.py [hunk 1]
```diff
@@ -108,14 +111,23 @@ def __post_init__(self):
 
         if self.envelope_type == 'gauss':
             self.equation = "exp(-2*ln(2)*(t-t0)^2/fwhm^2)*cos((omega+chirp*t)*t)"
+            extent = 2.4 * self.fwhm
```
Comment (danielhollas) [inline src/promdens/promdens.py:114]: Not sure about the name of this variable, maybe `temporal_extent`. Or something simpler?

**Comment by danielhollas** [comment]
I am having numerical issues with the new tests but apart from that this is ready for review @JanosJiri

File: src/promdens/promdens.py [hunk 2]
```diff
@@ -299,34 +322,24 @@ def calc_field(self, pulse: LaserPulse) -> None:
         :param pulse: LaserPulse dataclass containing laser pulse parameters (frequency, fwhm...)
         """
         self.pulse = pulse
+        # TODO: Remove all these extra assignments
```
Comment (danielhollas) [inline src/promdens/promdens.py:325]: Going to do this in follow-up PR to keep this one small.

File: tests/test_pulse.py [hunk 3]
```diff
@@ -68,3 +68,199 @@ def test_field_cos_with_chirp(make_pulse):
...
+    pulse = make_pulse(fwhm=fwhm, t0=t0, envelope_type=envelope_type)
+
+    t = np.array([-64, -16, -4.0, 0.0, 4.0, 16, 64])
+
+    envelope = pulse.calc_field_envelope(t)
+
+    s = snapshot(
+        {
+            "gauss": {
+                0: 1.0960549698939707e-11,
```
Comment (danielhollas) [inline tests/test_pulse.py:86]: We could save the reference values to a file, but I think it's quite useful to see them directly here.

File: tests/test_pulse.py [hunk 4]
```diff
@@ -68,3 +68,199 @@ def test_field_cos_with_chirp(make_pulse):
...
+                3: 1.0,
+                4: 0.90982893698841,
+                5: 0.11811460943703297,
+                6: 0.0,
+            },
+        }
+    )
+    s_tmin = snapshot(
+        {
+            "gauss": -36.0,
```
Comment (danielhollas) [inline tests/test_pulse.py:134]: @JanosJiri please double check that these values make sense. 

For example it's quite nice to see how much "wider" the lorentzian pulse is.

**Comment by danielhollas** [comment]
The tests are now passing, I've used [pytest.approx](https://docs.pytest.org/en/8.3.x/reference/reference.html#pytest-approx) for inexact comparisons. It's not a fully satisfying solution since it doesn't integrate as well with the `inline-snapshot` library. I'll keep mulling it over but it should not block this PR.

**Comment by JanosJiri** [comment]
> The tests are now passing, I've used [pytest.approx](https://docs.pytest.org/en/8.3.x/reference/reference.html#pytest-approx) for inexact comparisons. It's not a fully satisfying solution since it doesn't integrate as well with the `inline-snapshot` library. I'll keep mulling it over but it should not block this PR.

I quite like it. What I had a problem with when doing the tests was the absolute comparison and threshold for absolute differences. This allows us to compare relative differences, which might be very handy.

File: src/promdens/promdens.py (hunk 1 above)
Comment (JanosJiri) [reply src/promdens/promdens.py:114]: I would go with temporal_extent. It's long but it explains the meaning of the variable. I will change it in a next short pull request with some minor modifications.

File: src/promdens/promdens.py (hunk 2 above)
Comment (JanosJiri) [reply src/promdens/promdens.py:325]: Ok, cool.

File: tests/test_pulse.py (hunk 3 above)
Comment (JanosJiri) [reply tests/test_pulse.py:86]: I like it that way. Fewer files with data for tests are better, having everything in one place, especially for testing internal functions. Thanks!

File: tests/test_pulse.py (hunk 4 above)
Comment (JanosJiri) [reply tests/test_pulse.py:134]: I went through most of the data and it makes sense. I will change the `envelopes_analysis` and we will see if it still fulfils the FWHM conditions.

**Review by JanosJiri — APPROVED** [review]
Nice, thanks for the work on it. I like the current solution with `pytest.approx`. Might not be the best but better than what we had before. I check that the envelopes are still the same.

**Pull request merged**